SwiftCompilerSources/Sources/Optimizer/FunctionPasses/EffectInference.swift

```diff
@@ -0,0 +1,197 @@
+            return nil
+          }
+        })
+
+        // Update global effects
+        globalEffects.merge(callee.effects.globalEffects)
+
+        // Update argument effects
+        let arguments = apply.argumentOperands
+        for (paramIdx, effect) in sideeffects {
+          if let argIdx = apply.callerArgIndex(calleeArgIndex: paramIdx) {
+            updateEffect(to: arguments[argIdx].value, { $0.merge(effect) })
+          } else {
+            globalEffects.sideeffect.merge(effect)
+          }
+        }
+      }
+    case let fl as FixLifetimeInst:
+      return updateEffect(to: fl.operand, { $0.read = true })
+    case is AllocStackInst, is DeallocStackInst:
+      break
+    case let ba as BeginAccessInst:
+      if !ba.isStatic {
+        globalEffects.traps = true
+      }
+    case is EndAccessInst, is BeginBorrowInst, is EndBorrowInst:
+      break
+    case is AllocRefInst, is AllocRefDynamicInst:
+      globalEffects.allocates = true
+    case is StrongRetainInst, is RetainValueInst:
+      // TODO: strong_retain_unowned etc?
+      updateEffect(to: (inst as! UnaryInstruction).operand, {
+        $0.retain = true
+      })
+    case is StrongReleaseInst, is ReleaseValueInst:
+      updateEffect(to: (inst as! UnaryInstruction).operand, { $0.release = true })
+    case let cast as UnconditionalCheckedCastInst:
+      updateEffect(to: cast.operand, { $0.read = true })
+      globalEffects.traps = true
+    case let lb as LoadBorrowInst:
+      updateEffect(to: lb.operand, { $0.read = true })
+    case let load as LoadInst:
+      updateEffect(to: load.operand, {
+        $0.read = true
+        // TODO: Ownership Qualifier?
+      })
+    case let store as StoreInst:
+      updateEffect(to: store.destinationOperand.value, {
+        $0.write = true
+        if store.destinationOwnership == .assign {
+          $0.release = true
+        }
+      })
+    case is CondFailInst:
+      globalEffects.traps = true
+    case let pa as PartialApplyInst:
+      for (i, arg) in pa.arguments.enumerated() {
+        if pa.getArgumentConvention(calleeArgIndex: i).isIndirect {
+          updateEffect(to: arg) { $0.read = true }
+        }
+      }
+    case let bi as BuiltinInst:
+      switch bi.id {
+      case .IsUnique:
+        globalEffects.readsRC = true
+      case .CondUnreachable:
+        globalEffects.traps = true
+      default:
+        matched = false
+      }
+    default:
+      matched = false
+    }
+
+    if !matched {
+      globalEffects.sideeffect.read = globalEffects.sideeffect.read || inst.mayReadFromMemory
+      globalEffects.sideeffect.write = globalEffects.sideeffect.write || inst.mayWriteToMemory
+      if inst.mayHaveSideEffects {
+        globalEffects.setWorstEffects()
+      }
+
+      if inst.mayTrap {
+        globalEffects.traps = true
+      }
+    }
+  }
+
+    private mutating
+  func updateEffect(_ origin: Value, _ update: (inout SideEffect) -> (), _ path: SmallProjectionPath! = nil) {
```

Comment:
* Suggestion: `func updateEffect(forRoot value: Value,...`.

* If you make the effect flags a `let`, then this closure has to be `_ getEffects: () -> SideEffect`.

* `SmallProjectionPath! = nil` -> `SmallProjectionPath`

* nitpick: Also, please move this function below the other update function, because it's called from the other function.